fusion/test/TestBench.hpp

```diff
@@ -0,0 +1,254 @@
+    //! \brief top level extractor test runner
+    bool fieldExtractorTests(bool debug = false);
+
+    //! \brief create an inst from opcode, catch conversion errors
+    FieldExtractor::InstPtrType makeInst(MavisType &, uint32_t);
+
+    //! \brief helper for testing field values
+    bool testFieldValue(uint32_t, std::string, uint32_t act,uint32_t exp);
+
+    //! \brief transform funcs
+    static bool f1_constraints(FusionGroupType &, InstPtrListType &,
+                               InstPtrListType &);
+
+    //! \brief assign to InstPtrListType from opcodes
+    void assign(InstPtrListType &, std::vector<Opcode> &,
+                const FileNameListType &);
+    //! \brief info debug function
+    void info(InstUidListType &, InstUidListType &, InstPtrListType &);
+
+    //! \brief support golden reference hashes
+    void generateExpectHashes(
+        std::unordered_map<std::string, fusion::HashType> &,
+        const FusionType::FusionGroupCfgListType &);
+
+    //! \brief duplicate of hash function found in fusion group
+    //!
+    //! duplicated  to support debug
+    fusion::HashType jenkinsOneAtATime(const std::vector<fusion::UidType> &);
+
+    //! \brief common isa files to separate from the cmd line versions
+    static const std::vector<std::string> stdIsaFiles;
+
+    //! \brief extra messages
+    bool verbose{false};
+
+    //! \brief zoo.F1  example fusion group
+    //!
+    //! This is from the fusion group zoo, F1 has three variants. There are
+    //! three forms for each, opcode, uid, and asm text. I'm trying out
+    //! various methods for a user to express a fusion group when not using
+    //! the DSL. Note all of these will be needed.
+    //! @{
+    static InstUidListType uf1, uf1_1, uf1_2, uf1_3;
+    static OpcodeListType of1, of1_1, of1_2, of1_3;
+    //! @}
+
+    //! \brief zoo.F2 example fusion group
+    //! @{
+    static InstUidListType uf2;
+    static OpcodeListType of2;
+    //! @}
+
+    //! \brief zoo.F3 example fusion group
+    //! @{
+    static InstUidListType uf3;
+    static OpcodeListType of3;
+    //! }@
+
+    //! \brief zoo.F4 example fusion group
+    //! @{
+    static InstUidListType uf4;
+    static OpcodeListType of4;
+    //! }@
+
+    //! \brief zoo.F5 example fusion group and three variants
+    //! @{
+    static InstUidListType uf5, uf5_1, uf5_2, uf5_3;
+    static OpcodeListType of5, of5_1, of5_2, of5_3;
+    //! }@
+};
+
+// ---------------------------------------------------------------------
+//! \brief call back proxies used in the unit test(s)
+//!
+//! There is a callback for each fusion group test case  f1, f1.1, etc
+// ---------------------------------------------------------------------
+struct TestBench::cbProxy
+{
+    //! \brief ...
+    static bool uf1_func(FusionGroupType &, InstPtrListType &,
+                         InstPtrListType &)
+    {
+        cout << "HERE uf1_func called" << endl;
```

Comment:
shouldn't we be suing Msg class instead of couts?

Reply:
fixed